### ISSUE TITLE
`dtype_policy` doesn't function correctly when used in subclassing

### ISSUE THREAD

**Issue by james77777778** [keras-team/keras#19381]
This issue will lead to failure when using `quantize` in a subclassed model.

Here's an example:

```python
import tempfile

import keras
import keras.api_export
import keras.ops
import keras.random
import keras.saving


@keras.saving.register_keras_serializable("MyPackage")
class MySubclass(keras.layers.Layer):
    def __init__(self, dtype=None, **kwargs):
        super().__init__(**kwargs)
        self.layer1 = keras.layers.Dense(4, dtype=dtype)
        self.layer2 = keras.layers.BatchNormalization(axis=-1, dtype=dtype)
        self.layer3 = keras.layers.ReLU(dtype=dtype)

    def call(self, inputs):
        x = self.layer1(inputs)
        x = self.layer2(x)
        return self.layer3(x)

    def get_config(self):
        config = super().get_config()
        config.update({"dtype": self.dtype_policy})
        return config


model = keras.Sequential([keras.layers.Input([8]), MySubclass()])
model.quantize("int8")

print("dtype policy in MySubclass doesn't change after `quantize`")
print(model.dtype_policy)  # <FloatDTypePolicy "float32">

with tempfile.TemporaryDirectory() as tempdir:
    path = f"{tempdir}/model.keras"
    model.save(path)
    reloaded_model = keras.saving.load_model(path)  # <= failed

```

The error msg:

```bash
Layer 'dense_1' expected 2 variables, but received 3 variables during loading. Expected: ['kernel', 'bias']
```

The root cause is that `dense_1` should be configured as `"int8_from_float32"` to include an extra variable `kernel_scale`

My thought:
Perhaps we need a more fine-grained control over the dtype policy?
A verbose solution based on the current codebase might be as follows:

```python
@keras.saving.register_keras_serializable("MyPackage")
class MySubclass(keras.layers.Layer):
    def __init__(self, dtype=None, **kwargs):
        super().__init__(**kwargs)
        layer1_dtype = dtype["layer1"] if isinstance(dtype, dict) else dtype
        layer2_dtype = dtype["layer2"] if isinstance(dtype, dict) else dtype
        layer3_dtype = dtype["layer3"] if isinstance(dtype, dict) else dtype
        self.layer1 = keras.layers.Dense(4, dtype=layer1_dtype)
        self.layer2 = keras.layers.BatchNormalization(
            axis=-1, dtype=layer2_dtype
        )
        self.layer3 = keras.layers.ReLU(dtype=layer3_dtype)

    def call(self, inputs):
        x = self.layer1(inputs)
        x = self.layer2(x)
        return self.layer3(x)

    def get_config(self):
        config = super().get_config()
        config.update(
            {
                "dtype": {
                    "layer1": self.layer1.dtype_policy,
                    "layer2": self.layer2.dtype_policy,
                    "layer3": self.layer3.dtype_policy,
                }
            }
        )
        return config
```

This issue should impact subclassed layers in both KerasNLP and KerasCV

**Comment by sampathweb** [comment]
@james77777778 - Thanks for reporting this issue. Would you be able to self assign and work on this issue?

**Comment by james77777778** [comment]
> @james77777778 - Thanks for reporting this issue. Would you be able to self assign and work on this issue?

Hi @sampathweb 
I'm willing to resolve this issue but I think it requires a significant refactor of the current design.
We should wait for @fchollet 's judgement.

**Comment by fchollet** [comment]
I don't quite get it -- why wouldn't `quantize` just call `quantize` on all sublayers and thus set `dtype_policy`, including `dense_1`?

**Comment by james77777778** [comment]
> I don't quite get it -- why wouldn't `quantize` just call `quantize` on all sublayers and thus set `dtype_policy`, including `dense_1`?

The behavior of `quantize` is correct but the initialization of the subclass is incorrect when loading from the quantized model.

For example, in KerasNLP, `TransformerDecoder` uses `self.dtype_policy` to initialize sublayers:
https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/modeling/transformer_decoder.py#L134

If the sublayers like `self._feedforward_intermediate_dense` are quantized to `"int8_from_float32"`, this configuration won't be reflected when **loading** the model. This is because `TransformerDecoder` uses its own dtype policy (by default `"float32"`) for initializing all sublayers.

Colab to demonstrate:
https://colab.research.google.com/drive/1HJRj8RhSA9U1UmyVxwJLrGLU1Lj_gZRK?usp=sharing

**Comment by fchollet** [comment]
That's only true for subclassed layers that explicitly pass a `dtype` arg to their sublayers, right? If no arg is passed, what gets deserialized is the dtype associated with the config of each layer?

**Comment by james77777778** [comment]
> That's only true for subclassed layers that explicitly pass a `dtype` arg to their sublayers, right? If no arg is passed, what gets deserialized is the dtype associated with the config of each layer?

Actually, this is not true in the current codebase.

You can refer to the colab:
https://colab.research.google.com/drive/1vKLWhkOkz7Yb-Y1PlcEYIZqTIkIyOZgG?usp=sharing

```python
import tempfile

import keras
import keras.api_export
import keras.ops
import keras.random
import keras.saving


@keras.saving.register_keras_serializable("MyPackage")
class MySubclass(keras.layers.Layer):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.layer1 = keras.layers.Dense(4)  # no `dtype`
        self.layer2 = keras.layers.BatchNormalization(axis=-1)  # no `dtype`
        self.layer3 = keras.layers.ReLU()  # no `dtype`

    def call(self, inputs):
        x = self.layer1(inputs)
        x = self.layer2(x)
        return self.layer3(x)

    def get_config(self):
        config = super().get_config()
        return config


model = keras.Sequential([keras.layers.Input([8]), MySubclass()])
model.quantize("int8")

print("dtype policy in MySubclass doesn't change after `quantize`")
print(model.dtype_policy)  # <FloatDTypePolicy "float32">

with tempfile.TemporaryDirectory() as tempdir:
    path = f"{tempdir}/model.keras"
    model.save(path)
    reloaded_model = keras.saving.load_model(path)  # <= failed

```

In the subclass, there is no explicit `dtype`.
However, it will be set to `dtype_policies.get(None)` in `Operation.__init__` of `Layer`

A more fine-grained control over `dtype` (or `dtype_policy`) might be necessary for the introduced quantization methods (int8, fp8, etc)

**Comment by fchollet** [comment]
> However, it will be set to dtype_policies.get(None) in Operation.__init__ of Layer

Should we use some kind of scope to resolve the default dtype? So that child layers inherit their dtype from the parent layer.

It's easy to make it work for layers created in `build()` but it would be more challenging to make it work for layers created in `__init__()`. Still doable though.

An alternative is to fix the deserialization process.

**Comment by fchollet** [comment]
Basically we can make it so that:

- `__init__` is automatically wrapped in a method that can add custom logic
- In that wrapper, we extract the `dtype` arg
- If set explicitly, we open a "dtype scope" with that dtype
- If not set explicitly (None), we resolve the default dtype to the dtype from the scope (if any) or the global dtype policy otherwise

Same with `build()`.

Result: if a dtype is passed to any layer, it gets propagated to any sublayer created in `build()` and `__init__()`.

What do you think?

**Comment by james77777778** [comment]
> Basically we can make it so that:
> 
>     `__init__` is automatically wrapped in a method that can add custom logic
> 
>     In that wrapper, we extract the `dtype` arg
> 
>     If set explicitly, we open a "dtype scope" with that dtype
> 
>     If not set explicitly (None), we resolve the default dtype to the dtype from the scope (if any) or the global dtype policy otherwise
> 
> 
> Same with `build()`.
> 
> Result: if a dtype is passed to any layer, it gets propagated to any sublayer created in `build()` and `__init__()`.
> 
> What do you think?

This idea should be feasible for a single dtype policy, but how can we handle mixed dtype polices?
(EX: a model/subclass containing `float32` and `int8_from_float32`)

Should a dtype scope have a mapping between the layer and its dtype?

**Comment by fchollet** [comment]
How about this:

```python
@keras.saving.register_keras_serializable("MyPackage")
class MySubclass(keras.Layer):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.layer1 = keras.layers.Dense(4)
        self.layer2 = keras.layers.BatchNormalization(axis=-1)
        self.layer3 = keras.layers.ReLU()
        self.quantize_mode = None

    def call(self, inputs):
        x = self.layer1(inputs)
        x = self.layer2(x)
        return self.layer3(x)

    def quantize(self, mode):
        super().quantize(mode)
        self.quantize_mode = mode

    def get_config(self):
        config = super().get_config()
        config.update({"quantize_mode": self.quantize_mode})
        return config
   
    @classmethod
    def from_config(cls, config):
        quantize_mode = config.pop("quantize_mode", None)
        obj = super(MySubclass, cls).from_config(config)
        if quantize_mode:
            obj.quantize(quantize_mode)
        return obj
        
```

**Comment by james77777778** [comment]
> How about this:
> 

This won't work because:
1. `quantize` is only applied to the leaf nodes of the model. In this case, `Dense`, `BatchNormalization` and `ReLU` will invoke `quantize` but not `MySubclass`
2. `quantize` must be called on a built object

However, I think adding logic in serialization/deserialization might be a better solution than refactoring.
I can finally have a workaround for saving and reloading `MySubclass`

You can refer to the colab:
https://colab.research.google.com/drive/1W-eLDOo93mcQpB-6pXH4b9KWsMYN8kRz?usp=sharing

The following steps are needed:
- Add `build` for the class
- Store `quantization_mode` of leaf nodes in `get_config`
- Restore (and we might need to build) the layers that are quantized.

**Comment by james77777778** [comment]
We can now have a simplified version of the code after #19488 that introduces a setter of `dtype_policy`

```python
import tempfile

import keras
from keras import dtype_policies
import numpy as np


@keras.saving.register_keras_serializable("MyPackage")
class MySubclass(keras.Layer):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.layer1 = keras.layers.Dense(4)
        self.layer2 = keras.layers.BatchNormalization(axis=-1)
        self.layer3 = keras.layers.ReLU()

    def call(self, inputs):
        x = self.layer1(inputs)
        x = self.layer2(x)
        return self.layer3(x)

    def get_config(self):
        config = super().get_config()
        quantized_dtype_policies = {}
        for idx, layer in enumerate(
            self._flatten_layers(include_self=False, recursive=False)
        ):
            if isinstance(layer.dtype_policy, dtype_policies.QuantizedDTypePolicy):
                quantized_dtype_policies[str(idx)] = layer.dtype_policy
        if quantized_dtype_policies:
            config.update({"quantized_dtype_policies": quantized_dtype_policies})
        return config

    @classmethod
    def from_config(cls, config):
        quantized_dtype_policies = config.pop("quantized_dtype_policies", None)
        obj = super(MySubclass, cls).from_config(config)
        if quantized_dtype_policies:
            for idx, layer in enumerate(obj._flatten_layers(include_self=False, recursive=False)):
                if str(idx) in quantized_dtype_policies:
                    layer.dtype_policy = quantized_dtype_policies[str(idx)]
        return obj


model = keras.Sequential([keras.layers.Input([8]), MySubclass()])
model.quantize("int8")

with tempfile.TemporaryDirectory() as tempdir:
    path = f"{tempdir}/model.keras"
    model.save(path)
    reloaded_model = keras.saving.load_model(path)

x = keras.random.uniform([1, 8])
y1 = model(x, training=False)
y2 = reloaded_model(x, training=False)
np.testing.assert_allclose(y1, y2)
```

The main concept is to memorize and restore `dtype_policy` for each sublayer.

**Comment by james77777778** [comment]
This has been resolved now.
